File: allsample/src/main/java/com/allever/allsample/artistImageLoader/ImageResizer.java
package com.allever.allsample.artistImageLoader;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.util.Log;

import com.bumptech.glide.load.engine.Resource;

import java.io.FileDescriptor;

/**
 * Created by allever on 17-11-29.
 * @author Allever
 * 图片压缩
 */

public class ImageResizer {
    private static final String TAG = "ImageResizer";

    public ImageResizer(){}

    public Bitmap decodeSampleBitmapFromResource(Resources resources, int resId, int requestWidth, int requestHeight){
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeResource(resources,resId);
        options.inSampleSize = calculateSampleSize(options,requestWidth,requestHeight);
        options.inJustDecodeBounds = false;
        return BitmapFactory.decodeResource(resources,resId);
    }

    public Bitmap decodeSampleBitmpFromFileDescriptor(FileDescriptor fileDescriptor, int requestWidth, int requestHeight){
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeFileDescriptor(fileDescriptor,null,options);
        options.inSampleSize = calculateSampleSize(options, requestWidth,requestHeight);
        options.inJustDecodeBounds = true;
        return  BitmapFactory.decodeFileDescriptor(fileDescriptor, null, options);
    }

    private int calculateSampleSize(BitmapFactory.Options options, int requestWidth, int requestHeight){
        if (requestWidth == 0 || requestHeight == 0){
            return 1;
        }

        final int width = options.outWidth;
        final int height = options.outHeight;
        Log.d(TAG, "calculateSampleSize: outWidth = " + width + "\noutHeight = " + height);
        int inSampleSize = 1;
        Log.d(TAG, "calculateSampleSize: inSampleSize = " + inSampleSize);
        if (width > requestWidth || height > requestHeight){
            final int halfWidth = width/2;
            final int halfHeight = height/2;

            while (halfWidth/inSampleSize > requestWidth && halfHeight/inSampleSize > requestHeight){
//                inSampleSize = inSampleSize * 2;
                inSampleSize *= 2;
                Log.d(TAG, "calculateSampleSize: inSampleSize = " + inSampleSize);
            }
        }
        return inSampleSize;
    }


}

File: mysample/src/main/java/com/allever/mysimple/imageloader/lib/ImageResizer.kt
package com.allever.mysimple.imageloader.lib

import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import java.io.FileDescriptor

/***
 * 图片压缩类
 */
class ImageResizer() {


    fun decodeSampleFromResource(resources: Resources, resId: Int, reqWidth: Int, reqHeight: Int): Bitmap? {
        val option = BitmapFactory.Options()
        option.inJustDecodeBounds = true
        BitmapFactory.decodeResource(resources, resId, option)
        option.inSampleSize = calculateInSampleSize(option, reqWidth, reqHeight)
        option.inJustDecodeBounds = false
        return BitmapFactory.decodeResource(resources, resId, option)
    }

    fun decodeSampleFromFileDescroptor(fileDescriptor: FileDescriptor, reqWidth: Int, reqHeight: Int): Bitmap? {
        val option = BitmapFactory.Options()
        option.inJustDecodeBounds = true
        BitmapFactory.decodeFileDescriptor(fileDescriptor, null, option)
        option.inSampleSize = calculateInSampleSize(option, reqWidth, reqHeight)
        option.inJustDecodeBounds = false
        return BitmapFactory.decodeFileDescriptor(fileDescriptor, null, option)
    }


    /**
     * 计算采样率
     */
    private fun calculateInSampleSize(option: BitmapFactory.Options, reqWidth: Int, reqHeight: Int): Int {
        if (reqHeight == 0 || reqWidth == 0) {
            return 1
        }

        val width = option.outWidth
        val height = option.outHeight

        var inSampleSize = 1
        if (width > reqWidth || height > reqHeight) {
            val halfWidth = width / 2
            val halfHeight = height / 2
            while (halfWidth / inSampleSize >= reqWidth && halfHeight / inSampleSize >= reqHeight) {
                inSampleSize *= 2
            }
        }

        return inSampleSize
    }
}